ceno_zkvm/src/scheme/cpu/mod.rs

```diff
@@ -531,138 +522,66 @@ impl<E: ExtensionField, PCS: PolynomialCommitmentScheme<E>> TowerProver<CpuBacke
     {
         // First build tower witness
         let span = entered_span!("build_tower_witness", profiling_2 = true);
-        let (out_evals, prod_specs, logup_specs) =
+        let (mut out_evals, prod_specs, logup_specs) =
             self.build_tower_witness(composed_cs, input, records, is_padded, challenges);
         exit_span!(span);
 
         // Then prove the tower relation
         let span = entered_span!("prove_tower_relation", profiling_2 = true);
-        let res = CpuTowerProver::create_proof(out_evals, prod_specs, logup_specs, 2, transcript);
+        let (rt, proofs) = CpuTowerProver::create_proof(prod_specs, logup_specs, 2, transcript);
+
+        let lk_out_evals = out_evals.pop().unwrap();
+        let w_out_evals = out_evals.pop().unwrap();
+        let r_out_evals = out_evals.pop().unwrap();
         exit_span!(span);
-        res
+        (rt, proofs, lk_out_evals, w_out_evals, r_out_evals)
     }
 }
 
 impl<E: ExtensionField, PCS: PolynomialCommitmentScheme<E>> MainSumcheckProver<CpuBackend<E, PCS>>
     for CpuProver<CpuBackend<E, PCS>>
 {
     #[allow(clippy::type_complexity)]
-    #[tracing::instrument(
-        skip_all,
-        name = "build_main_witness",
-        fields(profiling_3),
-        level = "trace"
-    )]
-    fn build_main_witness<'a, 'b>(
+    #[tracing::instrument(skip_all, name = "table_witness", fields(profiling_3), level = "trace")]
```

Comment:
just an heads up: we should use tracing to wrap whole function for profiling.

to profile a fragment within a function, we could just wrap it e.g. 
```
let xxxx_span = entered_span!("xxxxx");
// code to profiling
exit_span!(xxxx_span)
```